home-work-function-pro.js

```diff
@@ -0,0 +1,188 @@
+  
+  // 4 Напишіть функцію detonatorTimer(delay), яка виводить в консоль число кожну секунду, починаючи з delay (ціле число) і в кінці замість 0 виведе 'BOOM!'
+
+  // Використовуючи setInterval
+
+  function detonatorTimer(delay) {
+    const countdown = delay;
+    const timerId = setInterval(() => {
+      if (countdown > 0) {
+        console.log(countdown);
+        countdown--;
+      } else {
+        console.log('BOOM!');
+        clearInterval(timerId);
+      }
+    }, 1000);
+  }
+  
+  console.log(detonatorTimer(3))
+
+  // Використовуючи вкладений setTimeout
+
+  function detonatorTimer(delay) {
+    let countdown = delay;
+  
+    function tick() {
+      if (countdown > 0) {
+        console.log(countdown);
+        countdown--;
+        setTimeout(tick, 1000);
+      } else {
+        console.log('BOOM!');
+      }
+    }
+  
+    setTimeout(tick, 1000);
+  }
+  
+  console.log(detonatorTimer(3))
+
+
+  // 5 Напишіть об'єкт в якому опишіть свої довільні властивості та довільні методи (2-3 штуки) що ці властивості виводять
+
+  const pet = {
+    name: 'Sirius',
+    species: 'Cat',
+    age: 1,
+    weight: 4,
+    mood: 'happy',
+    isHungry: false,
+    favoriteToys: 'ball',
+  
+    introduce() {
+      console.log(`My name is ${this.name}, I'm a ${this.species} and I'm ${this.age} years old`);
+    },
+    play() {
+      if (this.favoriteToys.length > 0) {
+        console.log(`I'm playing with my ${this.favoriteToys}`);
+      } else {
+        console.log(`I don't have any toys to play with`);
+      }
+    },
+    feed() {
+      if (this.isHungry == false) {
+        console.log(`Yum! That was delicious`);
+      } else {
+      console.log(`I need more chicken`);
+      }
+    },
+    checkMood() {
+      console.log(`I'm feeling ${this.mood} right now`);
+    }
+  };
+  
+  
+  pet.introduce(); // logs "My name is Sirius, I'm a Cat and I'm 1 years old" 
+  pet.play(); // logs "I'm playing with my ball" 
+  pet.feed(); // logs "Yum! That was delicious"
+  pet.checkMood(); // logs "I'm feeling happy right now" 
+  
+
+// 6 А тепер зробіть всі свої методи з задачі 5 прив'язаними до контексту свого об'єкту - аби вони були захищені від перезапису об'єкту і їх можна було викликати в таймері
+
+const securedPetIntroduce = pet.introduce.bind(pet);
+const securedPetPlay = pet.play.bind(pet);
+const securedPetFeed = pet.feed.bind(pet);
+const securedPetCheckMood = pet.checkMood.bind(pet)
+
+setTimeout(securedPetIntroduce, 1000); 
+setTimeout(securedPetPlay, 2000); 
+setTimeout(securedPetFeed, 3000); 
+setTimeout(securedPetCheckMood, 4000);
+
+// 7 Напишіть функцію-декоратор яка вопвільнює виконання довільної функції на вказану кількість секунд
+
+function someFunction(a, b) {
+  return a + b;
+}
+
+function slower(func, seconds) {
+  return function(...args) {
+    console.log(`You will see result in console in ${seconds} seconds`);
+    setTimeout(() => {
+      const result = func(...args);
+      console.log(result);
```

Comment:
you need to return the result of the `func` (use `call` or `apply` methods). `console.log` should be on line 188